src/main/java/org/apache/sling/xss/impl/CustomPolicy.java

```diff
@@ -0,0 +1,265 @@
+                boolean styleSeen = false;
+                // get the allowed Attributes for the tag
+                Map<String, Attribute> allowedAttributes = tag.getValue().getAttributeMap();
+                if (allowedAttributes != null && allowedAttributes.size() > 0) {
+
+                    // if there are allowed Attributes, map over them
+                    for (Attribute attribute : allowedAttributes.values()) {
+                        if (attribute.getOnInvalid().equals("removeTag")) {
+                            onInvalidRemoveTagList.add(attribute.getName());
+                        }
+                        if (CssValidator.STYLE_ATTRIBUTE_NAME.equals(attribute.getName()))
+                            styleSeen = true;
+                        List<String> allowedValuesFromAttribute = attribute.getLiterals();
+                        if (allowedValuesFromAttribute != null && allowedValuesFromAttribute.size() > 0) {
+                            for (String allowedValue : allowedValuesFromAttribute) {
+                                policyBuilder.allowAttributes(attribute.getName()).matching(true, allowedValue)
+                                        .onElements(tag.getValue().getName());
+                            }
+
+                        }
+                        List<Pattern> regexsFromAttribute = attribute.getPatternList();
+                        if (regexsFromAttribute != null && regexsFromAttribute.size() > 0) {
+                            policyBuilder.allowAttributes(attribute.getName())
+                                    .matching(matchesToPatterns(regexsFromAttribute))
+                                    .onElements(tag.getValue().getName());
+                        } else {
+                            policyBuilder.allowAttributes(attribute.getName()).onElements(tag.getValue().getName());
+                        }
+                    }
+
+                    if (!styleSeen) {
+                        policyBuilder.allowAttributes(CssValidator.STYLE_ATTRIBUTE_NAME)
+                                .matching(cssValidator.newCssAttributePolicy()).onElements(tag.getValue().getName());
+                    }
+                }
+                break;
+            }
+        }
+
+        // disallow style tag on specific elements
+        policyBuilder.disallowAttributes(CssValidator.STYLE_ATTRIBUTE_NAME)
+                .onElements(cssValidator.getDisallowedTagNames().toArray(new String[0]));
+
+        // ---------- dynamic attributes ------------
+        Map<String, Attribute> dynamicAttributes = new HashMap<>();
+        // checks if the dynamic attributes are allowed
+        if (policy.getDirectives().get("allowDynamicAttributes").equals("true")) {
+            dynamicAttributes.putAll(policy.getDynamicAttributes());
+            for (Attribute attribute : dynamicAttributes.values()) {
+                if (attribute.getOnInvalid().equals("removeTag")) {
```

Comment:
We should extract these string values as constants somewhere and reference them from the code.